### ISSUE TITLE
Project Lifecycle Review

### ISSUE THREAD

**Issue by knolleary** [flowforge/flowforge#392]
### Description

We need to review the overall lifecycle of a Project to cover - creation, stopping, restarting etc.

This directly feeds into the following stories:

- #348 
- #370 
- #377 
- #288

**Comment by hardillb** [comment]
State diagram as currently implemented (0.3.0)
```mermaid
stateDiagram-v2
SM: Safe Mode
Created
Starting
Running
Stopped
Crashed
state Crashed_Choice <<choice>>
%% Suspended
%% note right of Suspended
%%     Container stopped, but project 
%%     still in database
%% end note
Deleted
[*] --> Created
Created --> Starting
Starting --> Running
Running --> Stopped
Running --> Crashed
Crashed --> Crashed_Choice
Crashed_Choice --> SM: Loop detected
Crashed_Choice --> Starting: One off
SM --> Running
Stopped --> Starting
%% Stopped --> Suspended
Stopped --> Deleted
%% Suspended --> Deleted
%% Suspended --> Starting
Deleted --> [*]
```

A starting point to implement support for https://github.com/flowforge/flowforge/issues/377 and adding support to allow user to restart in safe mode

```mermaid
stateDiagram-v2
SM: Safe Mode
Created
Starting
Running
Stopped
Crashed
state Crashed_Choice <<choice>>
Suspended
note right of Suspended
    Container stopped, but project 
    still in database
end note
Deleted
[*] --> Created
Created --> Starting
Starting --> Running
Running --> Stopped
Running --> Crashed
Running --> SM
Crashed --> Crashed_Choice
Crashed_Choice --> SM: Loop detected
Crashed_Choice --> Starting: One off
SM --> Running
Stopped --> Starting
Stopped --> SM
Stopped --> Suspended
Stopped --> Deleted
Suspended --> Deleted
Suspended --> Starting
Deleted --> [*]
```

I think this would require a new container api endpoint of `suspend` and a requirement to pass arguments to `start` to allow safe mode and restarting from suspended state. 

The suspended state would also be the intermittent state when a stack is changed for a project, allowing the current container to be stopped and then restarted.

**Comment by knolleary** [comment]
Here's where we are headed. This visualisation separates the Container state (orange) and the Node-RED runtime state (red). Each transition is labelled with the API call that triggers a transition. It also shows the Project model's `state` value (grey) at each point.

This will drive some breaking changes to the existing container driver API - but will give us the level of control needed for the various stories outlined in this issue's description.

[figjam link](https://www.figma.com/file/uMLaDXcX7s50UabM035DqQ/Project-Lifecycle-Review) (haven't figured out how to automatically make links editable by people in the FF org, but read-only to anonymous.. so just ask if you want edit-access)

<img width="847" alt="image" src="https://user-images.githubusercontent.com/51083/161166428-cf95825b-d086-4ee8-aba4-b710b31832f9.png">


### Project States

The existing Project model has a `state` field that is used to store the user's desired state of the Project. This can be:
 - `started` - Container is running and Node-RED is running the flows
 - `stopped` - Container is running but Node-RED is stopped
 - `safe` - Container is running and Node-RED is running in Safe mode
 - `suspended` - Container is *not* running. This could be because the user has stopped it, or it failed to start. The `status()` api call (previously `details()`) can be used to get details.

When there is an internal transition between Node-RED states - such as if NR crashes in a loop and is placed into safe mode, the Project model isn't updated as the user's desired state would still be `started`.

### Container API

 - ~`create`~ - removed from the API. I couldn't see any work to be done between `create` and `start` to warrant the two separate functions.
 - `start` - starts the container (if it is not already running). This code should handle two cases:
     1. a container that has never been deployed yet - so there may be some runtime settings that are not known and need to be set
     2. a container that has been previously stopped - so some state may already exist in the runtime settings that should be restored
 - `stop` - stop the container running (eg, stop the nr-launcher process).
 - `remove` - delete the container and remove any resources it was using.
 - `startFlows` - requires the container to be started, will start Node-RED. Takes an optional arg on whether to start in safe mode
 - `stopFlows` - requires the container to be started, will stop Node-RED.
 - `status` - (previously called `details`) - returns the status information about the container, including the actual state.

### Scenarios

#### Start a Project

1. `project.state` set to `started` (the desired state)
2. call `container.start` - but do *not* wait for the promise to resolve. This allows us to respond promptly to the API request
    - if a call to `container.status` is made, the status should be `starting` until Node-RED is running (or near enough)
    - if the container fails to deploy, `project.state` is set to `suspended` and any call to `container.status` should return a state of `failed` (with some way of including any further message for the user)
3. the UI can then poll the status end-point for a project (TBD) to report back progress of starting.

#### Update Project Stack

1. `container.stop` - stop the container running
5. Update `project.stack` to point at the new stack
6. `container.start` - restart the container for the project, causing a rebuild on the new stack

#### Restart Node-RED to pickup change in runtime settings

1. `container.stopFlows`
2. `container.startFlows`

#### Suspend a Project

1. `container.stop`
2 ... that's all

**Comment by sammachin** [comment]
Is there a path where a project could go into suspended (container stopped) without user interaction? 
Or is it only automatically into safe mode but the container still runs?

Thinking about the impact on billing here, where we would probably not bill for suspended projects.

**Comment by knolleary** [comment]
You are right that the suspended project state has been added in consideration of #377, so we have a state where the project isn't consuming any resources and can be removed from billing whilst in that state.

The only identified trigger to put a project into that state is if the `container.stop` api is called. No transition has been identified where the project does that by itself. I cannot immediately think of a scenario that requires that, but of course, nothing precludes us from doing that in the future.
